keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,415 @@
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration,
+  denoting **desired** state of the Pod,
+* PodStatus.ContainerStatus.ResourceAllocated (new object) denotes **actual**
+  state of the Pod resources.
+
+In order to determine the state of a Pod resource update, we add two new
+PodConditions named PodResizeSuccess and PodResizePending.
+
+PodResizeSuccess PodCondition can have the following values:
+* Status: true  - Last resize request was successful,
+  - Reason: Success - Pod was resized successfully,
+* Status: false - Last resize failed,
+  - Reason: FailedNodeCapacity - Node does not have enough room,
+  - Reason: RejectedResourceQuota - Request exceeds ResourceQuota.
+  - Reason: RejectedQoSClassChange - Request attempts to change QoS class,
+  - Reason: Retrying - Controller initiated a retry.
+
+PodResizePending PodCondition can have the following values:
+* Status: false - PodResizeSuccess PodCondition shows status of last resize,
+  - Reason: Complete - Pod resize request handled to completion,
+* Status: true  - Resize request is being processed,
+  - Reason: AwaitingPreemption - Lower priority Pods being pre-empted,
+  - Reason: AwaitingPodResize - Kubelet is processing Pod resize.
+
+#### Container Restart Policy
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### Pod Resize Retry Policy
+
+If resource update fails, say due to lack of space on Node, default behavior
+is to let the initiating actor such as VPA handle the failure. Alternately, a
+Controller can either retry the update, or reschedule the pod based on policy.
+
+PodSpec is extended with a new flag, PodSpec.RetryPolicy, with possible values:
```

Comment:
rather than model it this way, should it be time bound?  i.e. if `DesiredResources` is false for a period of time, then the controller can do an action?

Reply:
I know k8s typically uses back-off and timed retries pattern. However, in this case, we felt that it is best to retry when other pods with the same or greater priority as the pod that failed resize leave the node. At this point, it has the greater chance of success - with any lower priority pods facing eviction at the point of retry.

Reply:
I am not sure this field as proposed is very clear.  `PodSpec.RetryPolicy` seems confusing with `PodSpec.RestartPolicy` .  This also seems to blend the concern between the controller (like the VPA with the pod spec).  Could the policy be captured on the VPA itself without changing the pod spec?  Why does the pod spec need to know anything about this?

Reply:
thinking about this more, I do not feel this belongs on the pod spec, and instead belongs on the controller/actor that is trying to drive the change.  Having a VerticalPodResizingPolicy on VPA is clearer than a RetryPolicy on PodSpec

Reply:
> thinking about this more, I do not feel this belongs on the pod spec, and instead belongs on the controller/actor that is trying to drive the change. Having a VerticalPodResizingPolicy on VPA is clearer than a RetryPolicy on PodSpec

Yes. This can be done in VPA. The default behavior for K8s in this case should be to keep retrying a failed update.

Reply:
I've discussed RetryPolicy with sig-autoscaling, and is out of this KEP, and we are going with resize backoff-retry typical of other k8s components. Resolving this.